### PR TITLE
Add proposal for AzureManagedCluster graduation from experimental

### PR DESCRIPTION
**Pull request by jackfrancis**
**What type of PR is this?**

/kind documentation

**What this PR does / why we need it**:

This PR is a proposal that outlines the set of criteria that must be addressed prior to graduating `AzureManagedCluster` out of experimental.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Contributes to #2204

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: docs/proposals/20220825-azuremanaged-cluster-v1.md [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+## Summary
+
+`AzureManagedCluster` is a capz-native implementation of Azure Managed Kubernetes (AKS). Because there is no standard set of Cluster API resource definitions for a "Managed Kubernetes cluster", it is left up to the provider to re-use the existing Cluster API specification (for example, the `Cluster` and its to-be-implemented-by-provider properties such as `ControlPlaneEndpoint`, `ControlPlaneRef` and `InfrastructureRef`). As a result, capz implemented `AzureManagedCluster` with an API contract designation of "experimental", to allow for rapid prototyping and discovery.
+
+With the recent adoption of `AzureManagedCluster` by the capz community for practical, real-world use, we want to identify the set of outstanding items that may prevent graduation to v1, and address each one of them, so that future adoption can be unlocked, and users can confidently build resilient systems on top of a stable API.
+
+## Motivation
+
+### Goals
+- Agree upon a definitive v1 specification of `AzureManagedCluster`
```
Comment (nojnhuh) [inline docs/proposals/20220825-azuremanaged-cluster-v1.md:29]: I imagine this would also include associated types like AzureManagedControlPlane and AzureManagedMachinePool?

File: docs/proposals/20220825-azuremanaged-cluster-v1.md [hunk 2]
```diff
@@ -0,0 +1,67 @@
+---
+title: AzureManagedCluster graduation to v1 from experimental
```
Comment (nojnhuh) [inline docs/proposals/20220825-azuremanaged-cluster-v1.md:2]: Does "v1" in this context refer to the literal `apiVersion` or more like "stable" in the sense that the current non-experimental v1beta1 APIs are stable?

**Review by nojnhuh — COMMENTED** [review]
Overall LGTM

I remember it being brought up before that there might be an implicit dependency here on graduating MachinePool out of experimental in CAPI before we can call AzureManagedCluster stable. Should we note in this doc that that's something we have yet to think through?

File: docs/proposals/20220825-azuremanaged-cluster-v1.md [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+creation-date: 2022-08-25
+last-updated: 2022-08-25
+status: implementable
+see-also:
+- https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/2204
+- https://github.com/kubernetes-sigs/cluster-api/pull/6988
+---
+
+
+# AzureManagedCluster graduation to v1 from experimental
```
Comment (CecileRobertMichon) [inline docs/proposals/20220825-azuremanaged-cluster-v1.md:18]: what do we mean by "v1" exactly? Are we talking about the API version (currently already v1beta, I'm not sure we want to graduate the API the v1 quite yet, as the rest of Cluster API CRDs are on v1beta1 still), or are we talking about removing the feature flag and moving the API to `api/` in the codebase instead of `exp/api`? I'm assuming it's the latter but would be good to be explicit, "v1" doesn't mean the same thing to everyone

File: docs/proposals/20220825-azuremanaged-cluster-v1.md [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+### Goals
+- Agree upon a definitive v1 specification of `AzureManagedCluster`
+- Prioritize timeliness of a v1 definition so that users can confidently commit to v1 in the near term
+- Prioritize "architectural affinity" with other Cluster API providers implementing Managed Kubernetes
+
+### Non-Goals / Future Work
+- Add features to `AzureManagedCluster`
+- Standardize any opinions about how best to use AKS
+- Promote capz-specific opinions about Managed Kubernetes across the Cluster API provider ecosystem
+- Include operational support in the scope of a v1 definition
```
Comment (CecileRobertMichon) [inline docs/proposals/20220825-azuremanaged-cluster-v1.md:37]: Again, what does this mean? Are we redefining what "v1" means and if so is it different from semver or API version v1 ?

**Review by CecileRobertMichon — COMMENTED** [review]
Would it make sense for this proposal to be an issue instead? Enhancement proposals are usually defining a new functionality that has associated user stories and implementation details. This is more of a task list/epic for graduating an existing functionality. On the other hand, `Integrate Cluster API Proposal Recommendations` is something we'd want to have a concrete proposal for, especially when it comes to user-facing API changes.

File: docs/proposals/20220825-azuremanaged-cluster-v1.md (hunk 1 above)
Comment (jackfrancis) [reply docs/proposals/20220825-azuremanaged-cluster-v1.md]: added this mention

File: docs/proposals/20220825-azuremanaged-cluster-v1.md (hunk 3 above)
Comment (jackfrancis) [reply docs/proposals/20220825-azuremanaged-cluster-v1.md]: I've replaced references to "v1" with "graduation from experimental". I agree that the term "v1" doesn't have much concrete meaning, and defining that (IMO) isn't part of this scope.

File: docs/proposals/20220825-azuremanaged-cluster-v1.md (hunk 4 above)
Comment (jackfrancis) [reply docs/proposals/20220825-azuremanaged-cluster-v1.md]: removed

**Comment by jackfrancis** [comment]
@CecileRobertMichon I agree that whatever happens we are going to want to file issues for commited work. Not sure if that can be helped by having a proposal PR that defines things in a markdown in the repo, or if we can simply "publish" the proposal in an issue description. We can discuss in tomorrow's office hours.

**Comment by CecileRobertMichon** [comment]
> We can discuss in tomorrow's office hours.

Did we end up discussing/coming to an agreement on this? personally I'm +1 on just keeping track of tasks as issues with a project board to track progress

**Comment by jackfrancis** [comment]
@CecileRobertMichon time didn't permit yesterday, but yeah I agree that this doc should pair with an "AzureManagedCluster graduation" project board.

I think I'd like to keep this doc open as a (IMO) preferable way to document to all AKS + capz stakeholders the agreed upon definition of "graduation".

@zmalik @sonasingh46 @mtougeron and others, please add any feedback that better reflects your ideas of desired progress, and share this with your CTOs :)

File: docs/proposals/20220825-azuremanaged-cluster-exp-graduation.md [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+### 4. Affinity between capz / Cluster API resource lifecycle enforcement and AKS lifecycle enforcement.
+
+Not unrelated to the above audit of AKS features, we should also overlay the Cluster API lifecycle combinatorics (simply speaking: Create, Read, Update, Delete) against the set of AKS cluster primitives to ensure that sane interfaces exist in capz in order to effectively enforce (for example, add an AKS node pool), or passively delegate authority (for example, defer to the built-in AKS autoscaler when enabled to enforce `MachinePool` replica count), if appropriate.
+
+### 5. Azure API Request Optimizations
+
+Prior to graduation from experimental, we should ensure that the sufficient set of configurable interfaces exist to allow "`AzureManagedCluster`" users to effectively tune capz so that no unnecessary Azure API requests are introduced into their AKS environment. We should ship sane, optimized defaults, but allow for flexible overrides to anticipate a wide variety of operational use cases.
```
Comment (mtougeron) [inline docs/proposals/20220825-azuremanaged-cluster-exp-graduation.md:63]: What do you mean by "unnecessary Azure API requests are introduced"? Are you referring to API requests made by capz during reconciliation or to those needed by an external client to configure the cluster?

File: docs/proposals/20220825-azuremanaged-cluster-exp-graduation.md [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+
+### 6. E2E Tests
+
+For every supported AKS feature, and for every supported lifecycle mutation of said feature, capz should have thorough, regular E2E test coverage.
+
+## Open Questions
+
+### Dependency upon (currently experimental) MachinePool spec
+
+At present we re-use the Cluster API `MachinePool` specification (as `AzureManagedMachinePool`) to implement AKS node pools running on Azure VMSS. A consideration here is that `MachinePool` is considered experimental, and behind a feature flag, by Cluster API. Do we want to add the graduation of `MachinePool` out of experimental as a prerequisite for graduating "`AzureManagedCluster`" out of experimental?
```
Comment (mtougeron) [inline docs/proposals/20220825-azuremanaged-cluster-exp-graduation.md:73]: Is there a proposal/kep that can be linked to for this? I think that'll make it easier for people to track the cross-project status.

**Review by mtougeron — COMMENTED** [review]
Just has 2 minor Qs but this looks great from my perspective.

File: docs/proposals/20220825-azuremanaged-cluster-exp-graduation.md (hunk 5 above)
Comment (jackfrancis) [reply docs/proposals/20220825-azuremanaged-cluster-exp-graduation.md:63]: Primarily I mean that capz + AKS offers an opportunity for more AKS API requests from two (that I can think of off the top of my head) dimensions:

1. the implementation details of eventual consistency introduce net new "let me check if the current state matches my goal state" GET calls that aren't there if you aren't running an eventual consistency controller on top of your AKS environment
2. if we do our jobs correctly capz will optimize for multi-cluster stories, which may yield customer scenarios where more clusters are running on single subscriptions, which is at least _n_x (where _n_ is the number of clusters) calls to the AKS API

So basically, we want to be a very well-behaved API citizen to introduce all of this neat foo while still keeping Azure subscriptions actually functional. :)

File: docs/proposals/20220825-azuremanaged-cluster-exp-graduation.md (hunk 6 above)
Comment (jackfrancis) [reply docs/proposals/20220825-azuremanaged-cluster-exp-graduation.md:73]: This is a good question. @mboersma @CecileRobertMichon is this the best current PR to reference for that effort?

https://github.com/kubernetes-sigs/cluster-api/pull/6089

File: docs/proposals/20220825-azuremanaged-cluster-exp-graduation.md [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+
+### 2. Integrate Cluster API Proposal Recommendations
+
+Once the proposal is accepted and merged into the Cluster API project as an endorsed set of provider recommendations, capz can audit the existing "`AzureManagedCluster`" experimental implementation for areas of disagreement with said recommendations. For each discovery of disagreement, there should be a defensible reason to matriculate a divergent capz Managed Kubernetes implementation into a post-experimental definition of "`AzureManagedCluster`", and ideally some form of sign-off from other provider maintainers. Where community consensus cannot be reached, capz should strongly consider evolving the experimental "`AzureManagedCluster`" implementation to meet the Cluster API Managed Kubernetes recommendation as a pre-requisite to graduating from experimental.
```
Comment (nojnhuh) [inline docs/proposals/20220825-azuremanaged-cluster-exp-graduation.md:56]: Regarding the possibility of a successor to the CAPI proposal, would we block CAPZ's non-experimental managed cluster implementation on that?

Comment (jackfrancis) [reply docs/proposals/20220825-azuremanaged-cluster-exp-graduation.md:56]: No. In fact there is no mention of a "Cluster API-native" ManagedCluster definition in this doc. Once a working group starts making progress towards this (assuming this proposal doc hasn't yet merged), it would probably make sense to mention that effort, and to explicitly state that AzureManagedCluster graduation is *not* dependent upon that proposal.

Off the top of my head I'd estimate that a long-term, capi ManagedCluster solution would be a `v1` thing (graduation of AzureManagedCluster will be within the `v1beta1` API scope).

File: docs/proposals/20220825-azuremanaged-cluster-exp-graduation.md (hunk 6 above)
Comment (jackfrancis) [reply docs/proposals/20220825-azuremanaged-cluster-exp-graduation.md:78]: added link

**Comment by jackfrancis** [comment]
@nojnhuh @CecileRobertMichon @mtougeron I've updated this proposal to reflect recent events, added some status to various items, and have started organizing links to issues that reflect work done. PTAL, I wonder if we can merge this soon and then regularly maintain it if things change.

**Comment by nojnhuh** [comment]
/lgtm
Thanks for putting this all together!

File: docs/proposals/20220825-azuremanaged-cluster-exp-graduation.md [hunk 8]
```diff
@@ -0,0 +1,126 @@
+---
+title: AzureManagedCluster graduation from experimental
+authors:
+  - "@jackfrancis"
+reviewers:
+  - @CecileRobertMichon
```
Comment (CecileRobertMichon) [inline docs/proposals/20220825-azuremanaged-cluster-exp-graduation.md:6]: did everyone listed in reviewers actually review? if not let's remove those who didn't prior to merge (or get their review)

**Review by CecileRobertMichon — COMMENTED** [review]
/lgtm

**Comment by nawazkh** [comment]
/lgtm
🚀

File: docs/proposals/20220825-azuremanaged-cluster-exp-graduation.md (hunk 8 above)
Comment (jackfrancis) [reply docs/proposals/20220825-azuremanaged-cluster-exp-graduation.md:6]: @zmalik @NovemberZulu @mtougeron could you kindly review this and weigh in?

If we don't hear from folks in ~a week or so let's remove those names from this list and finalize for merge.

Comment (mtougeron) [reply docs/proposals/20220825-azuremanaged-cluster-exp-graduation.md:6]: I'll take a look by EOD Monday.

File: docs/proposals/20220825-azuremanaged-cluster-exp-graduation.md [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+
+Once the proposal is accepted and merged into the Cluster API project as an endorsed set of provider recommendations, CAPZ can audit the existing "`AzureManagedCluster`" experimental implementation for areas of disagreement with said recommendations. For each discovery of disagreement, there should be a defensible reason to matriculate a divergent CAPZ Managed Kubernetes implementation into a post-experimental definition of "`AzureManagedCluster`", and ideally some form of sign-off from other provider maintainers. Where community consensus cannot be reached, CAPZ should strongly consider evolving the experimental "`AzureManagedCluster`" implementation to meet the Cluster API Managed Kubernetes recommendation as a pre-requisite to graduating from experimental.
+
+Ref:
+
+- https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2739
+- https://github.com/kubernetes-sigs/cluster-api/issues/7494
+
+
+### 3. Pathway Towards Full AKS Feature Support (Status: DONE)
```
Comment (jackfrancis) [inline docs/proposals/20220825-azuremanaged-cluster-exp-graduation.md:74]: @nojnhuh @mtougeron @sonasingh46 @zioproto @LochanRn @luthermonson 

All of you (and apologies to others I haven't mentioned!) have made significant feature contributions to the AzureManagedCluster API recently. As a result of that effort, does anyone have any concerns that the existing API surface area may be non-durable w/ respect to graduation from experimental?

The plan is to "move the existing API forward", based on the conclusion that we have done sufficient validation to confidently predict that additional feature work can continue to successfully land on this existing API. In other words, we have no reason to believe that we'll have to break the API in order to accommodate future AKS features.

Can folks weigh in on their confidence level of the above? Especially if you feel otherwise!

Thank you 🙏

Comment (nojnhuh) [reply docs/proposals/20220825-azuremanaged-cluster-exp-graduation.md:74]: One of the more potentially disruptive changes I can think of that we haven't addressed yet is exposing the AKS preview APIs (#2625). Even that seems like something we can enable in a backwards-compatible manner though. Overall, nothing else comes to mind that might threaten the integrity of the API based on where I've been focused recently.

**Comment by jackfrancis** [comment]
Lazy consensus timer starting, and expiring ~mid next week (week of Dec 19).

**Comment by jackfrancis** [comment]
Let's call the lazy consensus expired and land this officially before the winter break.

Thanks everyone for your feedback!

/approve
